### PR TITLE
=str #16986 Handle early cancelation issue

### PR DESCRIPTION
**Pull request by patriknw**
and =str #15191 Verify that stage actors are stopped

Some more things to verify and add, but you can take a first look and see if it is on right track. Note 2 commits if you want to review the two tickets in isolation.

### CODE REVIEW

**Comment by patriknw** [comment]
I think this PR is complete now. Please review.

Used checkThatAllStagesAreStopped in many tests.

Found one more bug of too early actor stop and thereby missing cancel of upstream , in fan-in tests ""work with one immediately failed and one nonempty publisher".

**Comment by ktoso** [comment]
LGTM

File: akka-stream-testkit/src/test/scala/akka/stream/testkit/StreamTestKit.scala [hunk 1]
```diff
@@ -174,4 +178,20 @@ object StreamTestKit {
...
+
+  def checkThatAllStagesAreStopped[T](block: ⇒ T)(implicit materializer: FlowMaterializer): T =
+    materializer match {
+      case impl: ActorFlowMaterializerImpl ⇒
+        impl.supervisor ! StreamSupervisor.StopChildren
+        val result = block
+        val probe = TestProbe()(impl.system)
+        probe.awaitAssert {
+          impl.supervisor.tell(StreamSupervisor.GetChildren, probe.ref)
+          val children = probe.expectMsgType[StreamSupervisor.Children].children
```
Comment (rkuhn) [inline akka-stream-testkit/src/test/scala/akka/stream/testkit/StreamTestKit.scala]: might want to set a small timeout on this per-iteration expectation

Comment (rkuhn) [inline akka-stream-testkit/src/test/scala/akka/stream/testkit/StreamTestKit.scala]: or not

**Comment by rkuhn** [comment]
LGTM! Will be fun to rebase TLS on top of that :-)

File: akka-stream-testkit/src/test/scala/akka/stream/testkit/StreamTestKit.scala (hunk 1 above)
Comment (patriknw) [inline akka-stream-testkit/src/test/scala/akka/stream/testkit/StreamTestKit.scala]: not, since I know that StreamSupervisor will reply immediately

Comment (rkuhn) [inline akka-stream-testkit/src/test/scala/akka/stream/testkit/StreamTestKit.scala]: yup, I figured ;-)

File: akka-stream/src/main/scala/akka/stream/impl/ActorFlowMaterializerImpl.scala [hunk 2]
```diff
@@ -222,6 +224,13 @@ private[akka] object StreamSupervisor {
   def props(settings: ActorFlowMaterializerSettings): Props = Props(new StreamSupervisor(settings))
 
   final case class Materialize(props: Props, name: String) extends DeadLetterSuppression
+
+  /** Testing purpose */
+  final case object GetChildren
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/ActorFlowMaterializerImpl.scala]: So the thing is, I believe we can make this a proper API. We need to discuss the details, I have some ideas. Good enough for this PR though, just wanted to mention.

File: akka-stream/src/main/scala/akka/stream/impl/Transfer.scala [hunk 3]
```diff
@@ -146,9 +154,31 @@ private[akka] trait Pump {
   private var currentAction: () ⇒ Unit =
     () ⇒ throw new IllegalStateException("Pump has been not initialized with a phase")
 
-  final def nextPhase(phase: TransferPhase): Unit = {
-    transferState = phase.precondition
-    currentAction = phase.action
+  final def initialPhase(waitForUpstream: Int, andThen: TransferPhase): Unit = {
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/Transfer.scala]: nice one!

**Comment by drewhk** [comment]
LGTM

**Comment by patriknw** [comment]
rebased, and fixed review comments, such as rename to `assertAllStagesStopped`

**Comment by patriknw** [comment]
real test failure that must be investigated before merging

**Comment by rkuhn** [comment]
real shame, I wanted to use this in the TLS tests ;-)

**Comment by patriknw** [comment]
I found the bug. You can review it in isolation in 8012cd4.
I will rebase.

**Comment by patriknw** [comment]
rebased again, and added one more test as requested by @drewhk : https://github.com/akka/akka/pull/17217/files#diff-59a559d890e855094590029696178df5R397

**Comment by patriknw** [comment]
We might have a problem with the TLS stage also as I mentioned to @rkuhn
I will not have time to try that today, so if this becomes green I will merge it.

**Comment by patriknw** [comment]
wat, did the target branch change again :(

**Comment by drewhk** [comment]
Lgtm
2015.04.23. 19:55 ezt írta ("Patrik Nordwall" <notifications@github.com>):

**Comment by patriknw** [comment]
the failures was in ConnectionPoolSpec, reported elsewhere

**Comment by patriknw** [comment]
Refs #16986 and #15191

**Pull request merged**